src/journal/catalog.c

```diff
@@ -94,25 +94,86 @@ const struct hash_ops catalog_hash_ops = {
         .compare = catalog_compare_func
 };
 
+static bool next_header(const char **s) {
+        const char *e;
+
+        e = strchr (*s, '\n');
+
+        /* Unexpected end */
+        if (!e)
+                return false;
+
+        /* End of headers */
+        if (e == *s)
+                return false;
+
+        *s = e + 1;
+        return true;
+}
+
+static char *combine_entries(const char *one, const char *two) {
```

Comment:
We generally prefer to return real errors, and pass back allocated pointers through a `**` pointer. This allows proper error propagation up the call stack.